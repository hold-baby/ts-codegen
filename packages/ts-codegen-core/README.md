# TS Codegen Core

[![Build Status](https://img.shields.io/travis/reeli/ts-codegen.svg?style=flat-square&branch=master)](https://travis-ci.org/reeli/ts-codegen)
[![codecov](https://codecov.io/gh/reeli/ts-codegen/branch/master/graph/badge.svg?style=flat-square)](https://codecov.io/gh/reeli/ts-codegen)
[![License](https://img.shields.io/npm/l/@ts-tool/ts-codegen.svg?style=flat-square)](https://npmjs.org/package/@ts-tool/ts-codegen)

Generate ts code from swagger.

# 开始

1. `npm install @ts-codegen-core`
