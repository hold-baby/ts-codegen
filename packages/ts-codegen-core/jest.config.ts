export default {
  testEnvironment: "node",
  modulePaths: ["<rootDir>"],
  moduleNameMapper: {
    "^@ts-tool/([^/]+)$": "<rootDir>/src/index.ts",
  },
  coverageDirectory: "coverage",
  testRegex: ".*/__tests__/.+\\.(generator|test|spec)\\.(ts|tsx)$",
  collectCoverage: true,
  collectCoverageFrom: ["src/**/*.{ts,tsx}", "!src/__types__/**", "!src/bootstrap.ts", "!src/index.ts"],
  transform: {
    "^.+\\.tsx?$": "ts-jest",
  },
};
