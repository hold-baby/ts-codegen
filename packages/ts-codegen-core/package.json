{
  "name": "@ts-tool/ts-codegen-core",
  "version": "3.3.0",
  "main": "dist/lib/index.js",
  "module": "dist/module/index.js",
  "types": "dist/index.d.ts",
  "files": [
    "dist",
    "README.md"
  ],
  "scripts": {
    "start": "ts-node -r tsconfig-paths/register src/bootstrap.ts"
  },
  "dependencies": {
    "axios": "0.21.x",
    "js-yaml": "4.0.x",
    "lodash": "4.17.x",
    "prettier": "*"
  },
  "publishConfig": {
    "access": "public"
  },
  "author": "reeli",
  "license": "ISC",
  "bugs": {
    "url": "https://github.com/reeli/ts-codegen/issues"
  },
  "homepage": "https://github.com/reeli/ts-codegen#readme"
}
