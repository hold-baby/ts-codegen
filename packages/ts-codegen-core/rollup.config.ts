import path from "path";
// @ts-ignore
import rollupBabel from "rollup-plugin-babel";
// @ts-ignore
import rollupTypeScript from "rollup-plugin-typescript";

const pkg = require(path.join(process.cwd(), "package.json"));

module.exports = {
  input: "src/index.ts",
  output: [
    {
      dir: "dist/lib",
      format: "cjs",
    },
    {
      dir: "dist/module",
      format: "es",
    },
  ],
  external: [
    // @ts-ignore
    ...Object.keys(process.binding("natives")),
    ...Object.keys(pkg.dependencies || {}),
    ...Object.keys(pkg.peerDependencies || {}),
  ],
  plugins: [
    rollupTypeScript({
      target: "es5",
      module: "es6",
    }),
    rollupBabel({
      plugins: ["babel-plugin-pure-calls-annotation"],
      exclude: "node_modules/**",
      extensions: [".js", ".jsx", ".ts", ".tsx"],
    }),
  ],
};
