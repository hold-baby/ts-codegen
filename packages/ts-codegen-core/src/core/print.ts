import { IClientConfig, RequestType, ScanOptions } from "../__types__/types";
import { IStore, DeclKinds } from "./createRegister";
import { prettifyCode, objToTypeStr } from "../utils/common";
import { sortBy, isEmpty, compact, keys, mapValues, isString } from "lodash";
import { CustomType } from "./Type";
import { DEFAULT_SERVICE_NAME_IN_HEADER } from "..";

export const printOutputs = (
  clientConfigs: IClientConfig[],
  decls: IStore["decls"],
  requestCreateMethod: string = "createRequest",
  options?: ScanOptions,
) => {
  return prettifyCode(`${printRequest(clientConfigs, requestCreateMethod, options)} \n\n ${printTypes(decls)}`);
};

const hasRequestBody = (TReq: IClientConfig["TReq"]) => TReq?.requestBody || (TReq || {})["requestBody?"];

const printRequest = (clientConfigs: IClientConfig[], requestCreateMethod: string, options?: ScanOptions): string => {
  const configs = sortBy(clientConfigs, (o) => o.operationId);

  return configs
    .map((v) => {
      const toUrl = () => `url: \`${v.url}\`,`;
      const toMethod = () => `method: "${v.method}",`;
      const toRequestBody = () => {
        if (!isEmpty(v.bodyParams)) {
          return `data: ${v.bodyParams!.length > 1 ? `{${v.bodyParams!.join(",")}}` : v.bodyParams![0]},`;
        }
        return v.contentType && hasRequestBody(v.TReq) ? "data: requestBody," : "";
      };
      const toQueryParams = () => {
        const params = toRequestParams(v.queryParams);
        return params ? `params: ${params},` : "";
      };
      const toHeaders = () => {
        if (v.contentType && options?.withServiceNameInHeader) {
          return `headers: { 'Content-Type': '${v.contentType}', '${
            isString(options?.withServiceNameInHeader)
              ? options?.withServiceNameInHeader
              : DEFAULT_SERVICE_NAME_IN_HEADER
          }': serviceName },`;
        }

        if (v.contentType) {
          return `headers: { 'Content-Type': '${v.contentType}' },`;
        }

        if (options?.withServiceNameInHeader) {
          return `headers: { '${
            isString(options?.withServiceNameInHeader)
              ? options?.withServiceNameInHeader
              : DEFAULT_SERVICE_NAME_IN_HEADER
          }': serviceName },`;
        }

        return "";
      };

      const toGenerators = () => {
        const TReq = generateTReq(v.TReq);
        const TResp = v.TResp?.toType(false);

        if (!TReq && !TResp) {
          return "";
        }
        if (!TResp) {
          return `<${TReq}>`;
        }
        return `<${TReq}, ${TResp}>`;
      };
      const toRequestInputs = () => {
        const getRequestBody = () => {
          if (!isEmpty(v.bodyParams)) {
            return v.bodyParams!;
          }
          return v.contentType && hasRequestBody(v.TReq) ? ["requestBody"] : "";
        };
        const list = compact([...v.pathParams, ...v.queryParams, ...getRequestBody()]);
        return isEmpty(list) ? "" : toRequestParams(list);
      };

      return `
${addComments(v, options?.withComments)}
export const ${v.operationId} = ${requestCreateMethod}${toGenerators()}("${
        v.operationId
      }", (${toRequestInputs()}) => ({${toUrl()}${toMethod()}${toRequestBody()}${toQueryParams()}${toHeaders()}})
);
`;
    })
    .join("");
};

const addComments = (v: IClientConfig, withComments?: boolean) => {
  const summaryComment = withComments && v.summary ? `* ${v.summary}` : "";
  const deprecatedComment = v.deprecated ? `* @deprecated ${v.operationId}` : "";
  const comments = [summaryComment, deprecatedComment].filter((c) => !!c);

  return isEmpty(comments)
    ? ""
    : `/**
  ${comments.join("\\n")}
  */`;
};

const printTypes = (decls: IStore["decls"]): string => {
  return keys(decls)
    .sort()
    .map((k) => {
      const expr = decls[k].kind === DeclKinds.type ? "=" : "";
      const semi = decls[k].kind === DeclKinds.type ? ";" : "";
      return `export ${decls[k].kind} ${decls[k].name} ${expr} ${decls[k].type.toType()}${semi}`;
    })
    .join("\n\n");
};

const generateTReq = (TReq: IClientConfig["TReq"]) => {
  if (isEmpty(TReq)) {
    return;
  }

  function gen(obj: IClientConfig["TReq"]): string {
    return objToTypeStr(
      mapValues(obj, (v) => {
        if (!v.toType) {
          return gen(v as RequestType);
        }
        return (v as CustomType).toType(false);
      }),
    );
  }

  return gen(TReq);
};

const toRequestParams = (data: string[]) => (!isEmpty(data) ? `{\n ${data.join(",\n")} \n}` : undefined);
