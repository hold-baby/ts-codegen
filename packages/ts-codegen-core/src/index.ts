export * from "./core";
export * from "./utils";
export * from "./constants";
export * from "./codegen";
export * from "./__types__/OpenAPI";
export * from "./__types__/types";
