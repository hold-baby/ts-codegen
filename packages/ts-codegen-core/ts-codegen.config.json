{
  "outputFolder": "clients",
  "requestCreateLib": "../examples/utils/createRequest",
  "requestCreateMethod": "createRequest",
  "apiSpecsPaths": [
    {
      "path": "https://petstore.swagger.io/v2/swagger.json",
      "name": "PetStoreService1"
    },
    {
      "path": "https://petstore.swagger.io/v2/swagger.yaml",
      "name": "PetStoreService2"
    },
    {
      "path": "./examples/swagger.json",
      "name": "SwaggerService"
    },
    {
      "path": "./examples/openapi.json",
      "name": "OpenApiService"
    },
    {
      "path": "./examples/openapi1.json",
      "name": "OpenApiService1"
    },
    {
      "path": "./examples/demo.yaml",
      "name": "DemoService"
    }
  ],
  "options": {
    "withComments": true,
    "typeWithPrefix": true,
    "backwardCompatible": true,
    "withServiceNameInHeader": true
  }
}
