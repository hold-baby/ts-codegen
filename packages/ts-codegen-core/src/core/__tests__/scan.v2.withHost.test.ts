import swaggerV3 from "../../../examples/openapi.json";
import { IOpenAPI, scan } from "@ts-tool/ts-codegen-core";
import { printOutputs } from "../print";

describe("swagger v2 with host", () => {
  it("should handle basic schemas correctly", () => {
    const { clientConfigs, decls } = scan(swaggerV3 as IOpenAPI);
    expect(
      printOutputs(clientConfigs, decls, "createRequestConfig", { withServiceNameInHeader: "Host" }),
    ).toMatchSnapshot();
  });
});
