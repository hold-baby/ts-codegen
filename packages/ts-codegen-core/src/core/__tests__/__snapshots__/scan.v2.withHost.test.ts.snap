// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`swagger v2 with host should handle basic schemas correctly 1`] = `
"export const addPet = createRequestConfig<
  {
    requestBody: NewPet;
  },
  Pet
>(\\"addPet\\", ({ requestBody }) => ({
  url: \`/api/v3/pets\`,
  method: \\"POST\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"application/json\\", Host: serviceName },
}));

export const createItem = createRequestConfig<
  {
    item?: Item;
  },
  Message
>(\\"createItem\\", ({ item }) => ({
  url: \`/api/v3/items\`,
  method: \\"POST\\",
  params: {
    item,
  },
  headers: { Host: serviceName },
}));

export const deletePet = createRequestConfig<{
  id: number;
}>(\\"deletePet\\", ({ id }) => ({ url: \`/api/v3/pets/\${id}\`, method: \\"DELETE\\", headers: { Host: serviceName } }));

export const findPetById = createRequestConfig<
  {
    id: number;
  },
  Pet
>(\\"findPetById\\", ({ id }) => ({ url: \`/api/v3/pets/\${id}\`, method: \\"GET\\", headers: { Host: serviceName } }));

export const findPets = createRequestConfig<
  {
    fee: { age?: number; text: string };
    limit?: number;
    status: (keyof typeof FindPetsStatus)[];
    tags?: (keyof typeof FindPetsTags)[];
  },
  Pet[]
>(\\"findPets\\", ({ tags, fee, limit, status }) => ({
  url: \`/api/v3/pets\`,
  method: \\"GET\\",
  params: {
    tags,
    fee,
    limit,
    status,
  },
  headers: { Host: serviceName },
}));

export const formMultipartWithFile = createRequestConfig<{
  requestBody: { data?: Pet; file: Blob; slice?: string[]; string?: string };
}>(\\"formMultipartWithFile\\", ({ requestBody }) => ({
  url: \`/api/v3/demo/forms/multipart\`,
  method: \\"POST\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"multipart/form-data\\", Host: serviceName },
}));

export const formMultipartWithFiles = createRequestConfig<{
  requestBody: { files: Blob[] };
}>(\\"formMultipartWithFiles\\", ({ requestBody }) => ({
  url: \`/api/v3/demo/forms/multipart-with-files\`,
  method: \\"POST\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"multipart/form-data\\", Host: serviceName },
}));

export const formUrlEncoded = createRequestConfig<{
  requestBody: RequestForm;
}>(\\"formUrlEncoded\\", ({ requestBody }) => ({
  url: \`/api/v3/demo/forms/url-encoded\`,
  method: \\"POST\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"application/x-www-form-urlencoded\\", Host: serviceName },
}));

export const generateFeedbackAndLeadReportUsingGet = createRequestConfig<undefined, Resource>(
  \\"generateFeedbackAndLeadReportUsingGet\\",
  () => ({ url: \`/api/v3/demo/feedback\`, method: \\"GET\\", headers: { Host: serviceName } }),
);

export const getItems = createRequestConfig<
  {
    limit?: number;
    page?: number;
    sort?: string[];
  },
  Items
>(\\"getItems\\", ({ page, limit, sort }) => ({
  url: \`/api/v3/items\`,
  method: \\"GET\\",
  params: {
    page,
    limit,
    sort,
  },
  headers: { Host: serviceName },
}));

export const postBody = createRequestConfig<{
  requestBody: Pet;
}>(\\"postBody\\", ({ requestBody }) => ({
  url: \`/api/v3/body\`,
  method: \\"POST\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"application/json\\", Host: serviceName },
}));

export const updatePet = createRequestConfig<{
  requestBody?: Cat | Dog;
}>(\\"updatePet\\", ({ requestBody }) => ({
  url: \`/api/v3/pets\`,
  method: \\"PATCH\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"application/json\\", Host: serviceName },
}));

export interface Cat extends Pet {
  age?: number;
  hunts?: boolean;
}

export type CatA = { country?: string; kind?: keyof typeof CatAKind };

export enum CatAKind {
  \\"Red Cat\\" = \\"Red Cat\\",
  \\"Blue Cat\\" = \\"Blue Cat\\",
}

export type CatB = { age?: boolean; kind?: keyof typeof CatBKind; name?: string };

export enum CatBKind {
  \\"Red Cat\\" = \\"Red Cat\\",
  \\"Blue Cat\\" = \\"Blue Cat\\",
}

export type CatC = Cat & string;

export type CatD = string & (Cat | Dog);

export type CatE = { country?: string };

export type CatF = Cat[] & Item[];

export interface Category {
  id?: number;
  name?: string;
}

export interface Dictionary {
  dictionary?: { [key: string]: { code?: number; text?: string } };
  dictionaryA?: { age: number; name: string; [key: string]: string | number };
  dictionaryB?: { age: string; name: string; [key: string]: string };
  dictionaryC?: { age?: number; name?: string; [key: string]: any };
  dictionaryD?: { [key: string]: any };
}

export interface Dog extends Pet {
  bark?: boolean;
  breed?: keyof typeof DogBreed;
}

export enum DogBreed {
  \\"Dingo\\" = \\"Dingo\\",
  \\"Husky\\" = \\"Husky\\",
  \\"Retriever\\" = \\"Retriever\\",
  \\"Shepherd\\" = \\"Shepherd\\",
}

export interface Error {
  code: number;
  message: string;
}

export enum FindPetsStatus {
  \\"available\\" = \\"available\\",
  \\"pending\\" = \\"pending\\",
  \\"sold\\" = \\"sold\\",
}

export enum FindPetsTags {
  \\"z\\" = \\"z\\",
  \\"b\\" = \\"b\\",
  \\"c\\" = \\"c\\",
  \\"a\\" = \\"a\\",
}

export interface Item {
  color?: string;
  id?: string;
  name?: string;
  price?: string;
}

export type Items = Item[];

export interface Message {
  message?: string;
}

export interface NewPet {
  contentType?: NewPetContentType;
  name: string;
  reason?: NewPetReason;
  tag?: string;
}

export type NewPetContentType = \\"01\\" | \\"02\\" | \\"03\\" | \\"04\\" | \\"05\\" | \\"06\\";

export type NewPetReason = \\"1\\" | \\"3\\" | \\"4\\" | \\"5\\" | \\"6\\";

export interface Pet extends NewPet {
  categories?: Category[];
  id: number;
}

export type Pets = Pet[];

export interface RequestForm {
  data: Pet;
  slice: string[];
  string: string;
}

export interface Resource {
  description?: string;
  filename?: string;
  open?: boolean;
  readable?: boolean;
}
"
`;
